ReSharper.FSharp/src/FSharp/FSharp.Psi.Daemon/src/Stages/PatternTypeHintsStage.fs

```diff
@@ -0,0 +1,257 @@
+        let result = collectTypeHintAnchorsForBinding binding
+        context.LocalNodes.AddRange(result)
+
+        x.VisitNode(binding, context)
+
+    override x.VisitTopBinding(binding, context) =
+        let result = collectTypeHintAnchorsForBinding binding
+        context.TopLevelNodes.AddRange(result)
+
+        x.VisitNode(binding, context)
+
+    override x.VisitMemberDeclaration(memberDecl, context) =
+        let result = collectTypeHintsAnchorsForMember memberDecl
+        context.TopLevelNodes.AddRange(result)
+
+        x.VisitNode(memberDecl, context)
+
+    override x.VisitPrimaryConstructorDeclaration(constructor, context) =
+        let result = collectTypeHintAnchorsForConstructor constructor
+        context.TopLevelNodes.AddRange(result)
+
+        x.VisitNode(constructor, context)
+
+    override x.VisitSecondaryConstructorDeclaration(constructor, context) =
+        let result = collectTypeHintAnchorsForConstructor constructor
+        context.TopLevelNodes.AddRange(result)
+
+        x.VisitNode(constructor, context)
+
+    override x.VisitForEachExpr(forEachExpr, context) =
+        let result = collectTypeHintAnchorsForEachExpr forEachExpr
+        context.LocalNodes.AddRange(result)
+
+        x.VisitNode(forEachExpr, context)
+
+type private PatternsHighlightingProcess(logger: ILogger, fsFile, settingsStore: IContextBoundSettingsStore,
+                                           daemonProcess: IDaemonProcess, settings) =
+    inherit FSharpDaemonStageProcessBase(fsFile, daemonProcess)
+
+    let createTypeHintHighlighting (fcsType: FSharpType) (displayContext: FSharpDisplayContext) range pushToHintMode =
+        let displayContext = displayContext.WithShortTypeNames(true)
+        TypeHintHighlighting(fcsType.Format(displayContext), range, pushToHintMode)
+
+    let getReturnTypeHint (decl: IParameterOwnerMemberDeclaration) pushToHintMode =
+        match decl with
+        | :? IConstructorDeclaration
+        | :? IAccessorDeclaration -> ValueNone
+        | _ ->
+
+        let equalsToken = decl.EqualsToken
+        let patterns = decl.ParameterPatterns
+        let range =
+            match decl with
+            | :? IBinding as binding when patterns.Count = 0 ->
+                binding.HeadPattern.GetDocumentRange().EndOffsetRange()
+
+            | :? IMemberDeclaration as memberDeclaration when patterns.Count = 0 ->
+                memberDeclaration.NameIdentifier.GetDocumentRange().EndOffsetRange()
+
+            | _ -> equalsToken.GetDocumentRange().StartOffsetRange()
+
+        let symbolUse =
+            match decl with
+            | :? IBinding as binding ->
+                let refPat = binding.HeadPattern.As<IReferencePat>()
+                if isNull refPat then Unchecked.defaultof<_> else
+                refPat.GetFcsSymbolUse()
+            | :? IMemberDeclaration as memberDeclaration ->
+                memberDeclaration.GetFcsSymbolUse()
+            | _ -> Unchecked.defaultof<_>
```

Comment:
Would getting it via `IFSharpDeclaration` work better?

Reply:
Oh, right now bindings are not an IFSharpDeclaration :(